### PR TITLE
New ContractKit

### LINKED ISSUE #360: Monorepo SNBAT build incompatible ABIs between procotol and sdk
### Expected Behavior

As a result of `yarn build`, the following should happen:
 1. Protocol Contracts are compiled using network `current`
 2. ABIs should be extracted from truffle artifacts
 3. Extracted ABIs should be used to generate ContractWrappers and ContractType in ContractKit
 4. ContractKit should be compiled (typescript)

ContractKit thus will break everytime ABI generated from protocol is not compatible with code present in SDK

Plus: Since CI does `build` the monorepo. Incompatibilities will break CI.

### Current Behavior

We do a similar thing but by environment, and we download artifacts from google-cloud instead of using locally generated ones.

### PR DESCRIPTION
**Pull request by mcortesi**
### Description

New contractkit.

Implemented features:

 * `yarn build` on protocol generates code for contractkit
 * cleanup protocol build
 * `yarn build` on root pkg breaks if contractkit doesn't compile with generated code
 * Fetch addresses from registry (only once)
 * Expose all web3 contract wrappers & cache them
 * Framework to create wrappers
 * Celo Wrappers for BondedDeposits, Exchange, Validators, GoldToken, StableToken
 * Contract kit is instantiable, can have many instances in one process
 * sendTransaction ported from walletkit

Missing: Test, complete Wrappers, validate API

### Tested

Not yet

### Other changes

 * Cleanup protocol build process.
 * Fix celo/utils npm publish
 * hoist ts-node to the root package

### Related issues

- Fixes #360 #359 #333 #257 

### Backwards compatibility

Doesn't break anything, it's a new package

### CODE REVIEW

File: packages/contractkit/src/index.ts [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+export { newGovernance } from './generated/Governance'
+export { newMultiSig } from './generated/MultiSig'
+export { newRandom } from './generated/Random'
+export { newRegistry } from './generated/Registry'
+export { newReserve } from './generated/Reserve'
+export { newSortedOracles } from './generated/SortedOracles'
+export { newStableToken } from './generated/StableToken'
+export { newValidators } from './generated/Validators'
+
+export function newWeb3(url: string) {
```
Comment (nambrot) [inline packages/contractkit/src/index.ts:6]: Is this function necessary?

File: packages/contractkit/src/utils/send-tx.ts [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+export interface TxOptions {
+  gasInflationFactor?: number
+  gasBeneficiary?: Address
+  gasCurrency?: Address
+  from?: Address
+  estimatedGas?: number | undefined
+}
+
+/**
+ * sendTransactionAsync mainly abstracts the sending of a transaction in a promise like
```
Comment (nambrot) [inline packages/contractkit/src/utils/send-tx.ts]: can call this `sendTransaction` now. Can also remove the TODO below and reference to sendTransactionFactory probably

File: packages/contractkit/src/utils/tx-result.ts [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+import { TransactionReceipt } from 'web3/types'
+
+const debug = debugFactory('contractkit:txresult')
+
+export function toTxResult(pe: PromiEvent<any>) {
+  return new TransactionResult(pe)
+}
+
+export class TransactionResult {
+  private hashFuture = new Future<string>()
```
Comment (nambrot) [inline packages/contractkit/src/utils/tx-result.ts]: Could you elaborate on why you chose Future's here over the build in promise? IMO the latter are more familiar to most developers?

File: packages/contractkit/src/wrappers/BondedDeposits.ts [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+  //     weight: await this.getAccountWeight(accountAddress),
+  //   }
+  // }
+
+  async getBondedDepositValue(account: string, noticePeriod: string) {
+    const deposit = await this.contract.methods.getBondedDeposit(account, noticePeriod).call()
+    return this.getValueFromDeposit(deposit)
+  }
+
+  async zipAccountTimesAndValuesToDeposits(
```
Comment (nambrot) [inline packages/contractkit/src/wrappers/BondedDeposits.ts]: What do you think about making some of these functions private?

File: packages/contractkit/src/wrappers/BaseWrapper.ts [hunk 5]
```diff
@@ -0,0 +1,32 @@
+import { ContractKit } from 'src/kit'
+import { TxOptions } from 'src/utils/send-tx'
+import { TransactionObject } from 'web3/eth/types'
+
+type Method<I extends any[], O> = (...args: I) => TransactionObject<O>
+
+export abstract class BaseWrapper<T> {
+  constructor(protected readonly kit: ContractKit, protected readonly contract: T) {}
+
+  protected proxySend<I extends any[], O>(methodFn: Method<I, O>) {
```
Comment (nambrot) [inline packages/contractkit/src/wrappers/BaseWrapper.ts:11]: Beautiful

File: packages/protocol/scripts/build.ts [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+  exec(`rm -rf "${ROOT_DIR}/typechain"`)
+
+  // const globPattern = `${BUILD_DIR}/contracts/@(${TruffleTestContracts.join('|')}).json`
+  const globPattern = `${BUILD_DIR}/contracts/*.json`
+  exec(
+    `yarn run --silent typechain --target=truffle --outDir "./types/typechain" "${globPattern}" `
+  )
+}
+
+function generateFilesForContractKit() {
```
Comment (nambrot) [inline packages/protocol/scripts/build.ts:102]: I would love for this functionality to be available in contractkit so that external developers can use this for their contracts!

File: packages/utils/src/collections.ts [hunk 7]
```diff
@@ -0,0 +1,9 @@
+export function zip<A, B, C>(fn: (a: A, b: B) => C, as: A[], bs: B[]) {
```
Comment (nambrot) [inline packages/utils/src/collections.ts:1]: What do you think about just using `lodash`s zip?

File: packages/contractkit/src/wrappers/BaseWrapper.ts (hunk 5 above)
Comment (mcortesi) [reply packages/contractkit/src/wrappers/BaseWrapper.ts:11]: hahaha I imaged you would like it ;)

File: packages/utils/src/collections.ts (hunk 7 above)
Comment (mcortesi) [reply packages/utils/src/collections.ts:1]: lodash is a pretty big library, if we only need just a couple of thing i woudl try to avoid it

File: packages/contractkit/src/wrappers/BondedDeposits.ts (hunk 4 above)
Comment (mcortesi) [reply packages/contractkit/src/wrappers/BondedDeposits.ts]: sure... this was just a port of what is already present in cli

File: packages/contractkit/src/utils/send-tx.ts (hunk 2 above)
Comment (mcortesi) [reply packages/contractkit/src/utils/send-tx.ts]: sure, forgot to do that when moving it

File: packages/contractkit/src/index.ts (hunk 1 above)
Comment (mcortesi) [reply packages/contractkit/src/index.ts:6]: not now... but if we end up wrapping web3 we'll probably want it

File: packages/utils/src/collections.ts (hunk 7 above)
Comment (nambrot) [reply packages/utils/src/collections.ts:1]: We already have it in `@celo/utils` :)

File: packages/celotool/package.json [hunk 8]
```diff
@@ -38,8 +38,7 @@
     "@types/web3": "^1.0.18",
     "@types/yargs": "^12.0.1",
     "jest": "^24.8.0",
-    "ts-jest": "^24.0.0",
-    "ts-node": "^7.0.1"
```
Comment (jmrossy) [inline packages/celotool/package.json:42]: why remove ts-node?

File: packages/contractkit/package.json [hunk 9]
```diff
@@ -0,0 +1,32 @@
+{
```
Comment (jmrossy) [inline packages/contractkit/package.json:1]: Now would be a good opportunity to specify a homepage property, maybe keywords too. For example:
https://github.com/babel/babel/blob/master/packages/babel-core/package.json#L7
https://www.npmjs.com/package/@babel/core

File: packages/contractkit/src/contract-cache.ts [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+  // private wrapperCache: Map<CeloContract, any> = new Map()
+  private wrapperCache: WrapperCacheMap = {}
+
+  constructor(readonly kit: ContractKit) {}
+
+  // getAttestations() {
+  //   return this.getWrapper(CeloContract.Attestations, newAttestations)
+  // }
+  getBondedDeposits() {
+    return this.getContract(CeloContract.BondedDeposits)
```
Comment (jmrossy) [inline packages/contractkit/src/contract-cache.ts:57]: Will you add types these? I see you set up the template below in `getContract` but aren't using it yet

File: packages/contractkit/src/web3-contract-cache.ts [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+import { newMultiSig } from 'src/generated/MultiSig'
+import { newRandom } from 'src/generated/Random'
+import { newRegistry } from 'src/generated/Registry'
+import { newReserve } from 'src/generated/Reserve'
+import { newSortedOracles } from 'src/generated/SortedOracles'
+import { newStableToken } from 'src/generated/StableToken'
+import { newValidators } from 'src/generated/Validators'
+import { ContractKit } from 'src/kit'
+
+const ContractFactories = {
```
Comment (jmrossy) [inline packages/contractkit/src/web3-contract-cache.ts:19]: what's the difference btwn this file and `contract-cache`?

File: packages/utils/package.json [hunk 12]
```diff
@@ -1,12 +1,13 @@
...
   "name": "@celo/utils",
-  "version": "0.0.5",
+  "version": "0.0.6-beta",
   "description": "Celo common utils",
   "author": "Celo",
   "license": "Apache-2.0",
   "main": "./lib/index.js",
   "types": "./lib/index.d.ts",
   "scripts": {
+    "prepublishOnly": "yarn build",
```
Comment (jmrossy) [inline packages/utils/package.json:10]: Why `Only` here?

File: packages/utils/src/bn.ts [hunk 13]
```diff
@@ -0,0 +1,11 @@
+import BN from 'bn.js'
+
+export function compareBN(a: BN, b: BN) {
```
Comment (jmrossy) [inline packages/utils/src/bn.ts:3]: We generally use BigNumber instead of BN.js whenever we can

File: packages/celotool/package.json (hunk 8 above)
Comment (mcortesi) [reply packages/celotool/package.json:42]: I moved everything to the root package.

For devDependencies that makes sense. It's easier to manage, and to have the same version across every project

File: packages/contractkit/src/contract-cache.ts (hunk 10 above)
Comment (mcortesi) [reply packages/contractkit/src/contract-cache.ts:57]: do you refer to specify the return types explicitly?
All methods are typed, and have concrete types, not `any` or `ts-ignores`

File: packages/contractkit/src/web3-contract-cache.ts (hunk 11 above)
Comment (mcortesi) [reply packages/contractkit/src/web3-contract-cache.ts:19]: We have 2 types of contract objects: web3 wrappers, and our own wrappers.

As discussed in the sdk meeting, we want to favor ours, and leave others accessible but not on the front seat.

File: packages/utils/package.json (hunk 12 above)
Comment (mcortesi) [reply packages/utils/package.json:10]: From the doc:
> prepublish: Run BEFORE the package is packed and published, as well as on local npm install without any arguments

[source](https://docs.npmjs.com/misc/scripts)

That's actually confusing and `npm` added the `Only` variant, to run **only when publishing**; which is what i wanted.

File: packages/utils/src/bn.ts (hunk 13 above)
Comment (mcortesi) [reply packages/utils/src/bn.ts:3]: Yeah, I haven't brought up that discussion yet.
web3 defaults to BN; to use BigNumber we need to purposely set it with each new contract instance. Which if we don't do can result in really confusing errors.

So i wanted to bring this up, and see if we want to review our library of choosing.

File: packages/contractkit/src/address-registry.ts [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+import { newRegistry } from 'src/generated/Registry'
+import { Registry } from 'src/generated/types/Registry'
+import { ContractKit } from 'src/kit'
+
+// Registry contract is always predeployed to this address
+const REGISTRY_CONTRACT_ADDRESS = '0x000000000000000000000000000000000000ce10'
+
+export class AddressRegistry {
+  private readonly registry: Registry
+  private readonly cache: Map<CeloContract, Address> = new Map()
```
Comment (asaj) [inline packages/contractkit/src/address-registry.ts:11]: Probably not something for v0, but what happens when this cache becomes outdated?

File: packages/contractkit/src/kit.ts [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+      gasInflationFactor: 1.3,
+    }
+
+    this.registry = new AddressRegistry(this)
+    this._web3Contracts = new Web3ContractCache(this)
+    this.contracts = new WrapperCache(this)
+  }
+
+  async setGasCurrency(token: CeloToken) {
+    this.defaultOptions.gasCurrency = await this.registry.addressFor(token)
```
Comment (asaj) [inline packages/contractkit/src/kit.ts]: One thing to note here, it's better to leave `gasCurrency` as null when using `Gold`.  In fact, I'm not sure that it will work if you actually specify the `GoldToken` address

File: packages/contractkit/src/kit.ts [hunk 16]
```diff
@@ -0,0 +1,68 @@
...
+  setGasCurrencyAddress(address: Address) {
+    this.defaultOptions.gasCurrency = address
+  }
+
+  sendTransaction(tx: Tx) {
+    const promiEvent = this.web3.eth.sendTransaction({
+      // @ts-ignore
+      gasCurrency: this.defaultOptions.gasCurrency,
+      // @ts-ignore
+      gasBeneficiary: this.defaultOptions.gasBeneficiary,
```
Comment (asaj) [inline packages/contractkit/src/kit.ts]: Did you mean `gasFeeRecipient`? I'm not sure that we want to be setting this manually, it's best to let `celo-blockchain` set it for us

File: packages/contractkit/src/kit.ts [hunk 17]
```diff
@@ -0,0 +1,68 @@
...
+  get defaultAccount() {
+    return this.web3.eth.defaultAccount
+  }
+
+  setGasCurrencyAddress(address: Address) {
+    this.defaultOptions.gasCurrency = address
+  }
+
+  sendTransaction(tx: Tx) {
+    const promiEvent = this.web3.eth.sendTransaction({
```
Comment (asaj) [inline packages/contractkit/src/kit.ts:52]: There's a hack around `gasPrice` that we should do. Basically, `web3` will try to set a default gas price for us, for which it uses `5 * minimum_gas_price_in_gold`, no matter what the `gasCurrency` is.

If we want to suggest a good gas price when we pay in other currencies we need to set `gasPrice: '0'`.  This will prevent `web3` from setting a default gas price and allow `celo-blockchain` to set the gas price, for which it will consider the `gasCurrency`

File: packages/contractkit/src/utils/send-tx.ts [hunk 18]
```diff
@@ -0,0 +1,50 @@
...
+  gasCurrency?: Address
+  from?: Address
+  estimatedGas?: number | undefined
+}
+
+/**
+ * sendTransaction mainly abstracts the sending of a transaction in a promise like
+ * interface.
+ */
+export async function sendTransaction<T>(
```
Comment (asaj) [inline packages/contractkit/src/utils/send-tx.ts:20]: What is the difference between this and `kit.sendTransaction`/`kit.sendTransactionObject`

File: packages/contractkit/src/utils/tx-result.ts [hunk 19]
```diff
@@ -0,0 +1,59 @@
...
+  constructor(pe: PromiEvent<any>) {
+    pe.on('transactionHash', (hash: string) => {
+      debug('hash: %s', hash)
+      this.hashFuture.resolve(hash)
+    })
+      .on('receipt', (receipt: TransactionReceipt) => {
+        debug('receipt: %O', receipt)
+        this.receiptFuture.resolve(receipt)
+      })
+      .on('confirmation', (confirmationNumber: number) => {
```
Comment (asaj) [inline packages/contractkit/src/utils/tx-result.ts]: If we have finality within 1 block, do we need this? Or is it sufficient to be able to wait for the receipt?

File: packages/contractkit/src/wrappers/BondedDeposits.ts [hunk 20]
```diff
@@ -0,0 +1,133 @@
...
+  }
+
+  // FIXME this.contract.methods.delegateRewards does not exist
+  async delegateRewardsTx(account: string, delegate: string) {
+    const sig = await this.getParsedSignatureOfAddress(account, delegate)
+
+    return this.contract.methods.delegateRole(roles.rewards, delegate, sig.v, sig.r, sig.s)
+  }
+
+  async getParsedSignatureOfAddress(address: string, signer: string) {
```
Comment (asaj) [inline packages/contractkit/src/wrappers/BondedDeposits.ts]: Should this be private?

File: packages/contractkit/src/wrappers/GoldTokenWrapper.ts [hunk 21]
```diff
@@ -0,0 +1,15 @@
+import { GoldToken } from 'src/generated/types/GoldToken'
+import { BaseWrapper } from 'src/wrappers/BaseWrapper'
+
+export class GoldTokenWrapper extends BaseWrapper<GoldToken> {
+  allowance = this.proxyCall(this.contract.methods.allowance)
+  balanceOf = this.proxyCall(this.contract.methods.balanceOf)
```
Comment (asaj) [inline packages/contractkit/src/wrappers/GoldTokenWrapper.ts]: There's possibly a bug with `GoldToken` where this always returns `2^256 - 1`. Can you check whether or not this is the case, and if it is, have this call `web3.eth.getBalance` instead?

File: packages/contractkit/src/wrappers/Validators.ts [hunk 22]
```diff
@@ -0,0 +1,145 @@
...
+
+    const { lesser, greater } = await this.findLesserAndGreaterAfterVote(
+      validatorGroup,
+      votingDetails.weight
+    )
+
+    return this.contract.methods.vote(validatorGroup, lesser, greater)
+  }
+
+  private async findLesserAndGreaterAfterVote(
```
Comment (asaj) [inline packages/contractkit/src/wrappers/Validators.ts:116]: Probably for a future PR, but I would imagine a more general version of this function will be useful for other places we use `SortedLinkedLists`

File: packages/contractkit/test/driver.test.ts [hunk 23]
```diff
@@ -0,0 +1,59 @@
+import { Address } from '../src/base'
+import { newKit } from '../src/kit'
+
+test('Test driver', async () => {
+  // Integration
+  // const kit = newKit('http://35.247.98.50:8545')
+  const kit = newKit('http://35.247.74.27:8545')
```
Comment (asaj) [inline packages/contractkit/test/driver.test.ts]: What is this? How do you intend on testing this going forward?

File: packages/utils/src/bn.ts (hunk 13 above)
Comment (asaj) [reply packages/utils/src/bn.ts:3]: +1 on having that conversation, but can we decouple it from this PR and use `BigNumber` for now?

**Review by asaj — COMMENTED** [review]
👏 👏 👏

File: packages/contractkit/src/kit.ts (hunk 15 above)
Comment (mcortesi) [reply packages/contractkit/src/kit.ts]: good to know

File: packages/contractkit/src/utils/tx-result.ts (hunk 19 above)
Comment (mcortesi) [reply packages/contractkit/src/utils/tx-result.ts]: good question...
this was ported from old contractkit..

i don't know if we want this or not

File: packages/contractkit/src/kit.ts (hunk 17 above)
Comment (mcortesi) [reply packages/contractkit/src/kit.ts:52]: what happens if it's a locally signed tx?
I imagine that we need to set the gasPrice there, since celo-blockchain won't be able to modify the tx

File: packages/contractkit/src/utils/send-tx.ts (hunk 18 above)
Comment (mcortesi) [reply packages/contractkit/src/utils/send-tx.ts:20]: this should be addressed later..
one is for `raw` sendTransaction and the other to sendTx from congtract wrappers

File: packages/contractkit/src/wrappers/Validators.ts (hunk 22 above)
Comment (mcortesi) [reply packages/contractkit/src/wrappers/Validators.ts:116]: sure, the moment we add other wrappers this will make sense

File: packages/contractkit/src/wrappers/GoldTokenWrapper.ts (hunk 21 above)
Comment (mcortesi) [reply packages/contractkit/src/wrappers/GoldTokenWrapper.ts]: is there an issue for this?

File: packages/contractkit/src/wrappers/BondedDeposits.ts (hunk 20 above)
Comment (mcortesi) [reply packages/contractkit/src/wrappers/BondedDeposits.ts]: yes

File: packages/contractkit/src/wrappers/GoldTokenWrapper.ts (hunk 21 above)
Comment (mcortesi) [reply packages/contractkit/src/wrappers/GoldTokenWrapper.ts]: I will use `web3.eth.getBalance` just in case

File: packages/contractkit/test/driver.test.ts (hunk 23 above)
Comment (mcortesi) [reply packages/contractkit/test/driver.test.ts]: the file is just an example, not a test...

For testing, use ganache and start from a snapshot that already has all contracts

Comment (mcortesi) [reply packages/contractkit/test/driver.test.ts]: See https://github.com/celo-org/celo-monorepo/issues/472

File: packages/contractkit/src/utils/tx-result.ts (hunk 19 above)
Comment (mcortesi) [reply packages/contractkit/src/utils/tx-result.ts]: I'll remove this

File: packages/contractkit/src/kit.ts (hunk 16 above)
Comment (mcortesi) [reply packages/contractkit/src/kit.ts]: what about locally signed tx?

File: packages/contractkit/src/address-registry.ts (hunk 14 above)
Comment (mcortesi) [reply packages/contractkit/src/address-registry.ts:11]: can this happen? I mean, proxies won't change

File: packages/utils/src/bn.ts (hunk 13 above)
Comment (mcortesi) [reply packages/utils/src/bn.ts:3]: I'll leave this for now (the code is inherited from cli) and focus on other things.
It will be addressed in a later PR

File: packages/contractkit/src/kit.ts (hunk 17 above)
Comment (ashishb) [reply packages/contractkit/src/kit.ts:52]: > I imagine that we need to set the gasPrice there, since celo-blockchain won't be able to modify the tx

Yes

File: packages/contractkit/src/utils/send-tx.ts [hunk 24]
```diff
@@ -0,0 +1,50 @@
...
+  const promiEvent = tx.send({
+    from: txOptions.from,
+    // @ts-ignore
+    gasCurrency: txOptions.gasCurrency,
+    // @ts-ignore
+    gasFeeRecipient: txOptions.gasFeeRecipient,
+    gas,
+    // Hack to prevent web3 from adding the suggested gold gas price, allowing geth to add
+    // the suggested price in the selected gasCurrency.
+    gasPrice: '0',
```
Comment (ashishb) [inline packages/contractkit/src/utils/send-tx.ts:48]: Add a comment that  this won't work with locally signed transactions.

File: packages/contractkit/src/kit.ts [hunk 25]
```diff
@@ -0,0 +1,70 @@
...
+  async setGasCurrency(token: CeloToken) {
+    this.defaultOptions.gasCurrency =
+      token === CeloContract.GoldToken ? undefined : await this.registry.addressFor(token)
+  }
+
+  set defaultAccount(address: Address) {
+    this.web3.eth.defaultAccount = address
+  }
+
+  get defaultAccount() {
```
Comment (ashishb) [inline packages/contractkit/src/kit.ts]: Nit: consider adding return type

File: packages/contractkit/src/kit.ts [hunk 26]
```diff
@@ -0,0 +1,70 @@
...
+    return this.web3.eth.defaultAccount
+  }
+
+  setGasCurrencyAddress(address: Address) {
+    this.defaultOptions.gasCurrency = address
+  }
+
+  sendTransaction(tx: Tx) {
```
Comment (ashishb) [inline packages/contractkit/src/kit.ts]: Nit: consider adding return type

File: packages/contractkit/src/kit.ts [hunk 27]
```diff
@@ -0,0 +1,70 @@
...
+      // @ts-ignore
+      gasCurrency: this.defaultOptions.gasCurrency,
+      // @ts-ignore
+      gasFeeRecipient: this.defaultOptions.gasFeeRecipient,
+      ...tx,
+    })
+    return toTxResult(promiEvent)
+  }
+
+  sendTransactionObject(txObj: TransactionObject<any>, options?: TxOptions) {
```
Comment (ashishb) [inline packages/contractkit/src/kit.ts]: Nit: consider adding return type

File: packages/contractkit/src/wrappers/Validators.ts [hunk 28]
```diff
@@ -0,0 +1,145 @@
...
```
Comment (ashishb) [inline packages/contractkit/src/wrappers/Validators.ts:116]: Nit: consider adding return type

File: packages/contractkit/test/driver.test.ts [hunk 29]
```diff
@@ -0,0 +1,59 @@
...
+  const account: Address = '0xCEa3eF8e187490A9d85A1849D98412E5D27D1Bb3'
+  // const account2: Address = '0xCEa3eF5e187490A9d85A1849D98412E5D27D1Bb3'
+
+  // To interact with contract we obtain them from kit
+  const goldToken = await kit.contracts.getGoldToken()
+
+  const balance = await goldToken.balanceOf(account)
+  console.log(`balance:`, balance)
+
+  // // internally kit does:
```
Comment (ashishb) [inline packages/contractkit/test/driver.test.ts]: Why is this commented out?

File: packages/contractkit/test/driver.test.ts (hunk 23 above)
Comment (ashishb) [inline packages/contractkit/test/driver.test.ts]: Why not read the addresses the wallet kit does?
https://github.com/celo-org/celo-monorepo/blob/master/packages/walletkit/test/utils.ts#L78

File: packages/dappkit/package.json [hunk 30]
```diff
@@ -6,7 +6,7 @@
   },
   "dependencies": {
     "@celo/walletkit": "0.0.4",
-    "@celo/utils": "0.0.5",
+    "@celo/utils": "0.0.6-beta",
```
Comment (ashishb) [inline packages/dappkit/package.json]: This is unrelated. Right?
Please send a separate PR.

File: packages/mobile/package.json [hunk 31]
```diff
@@ -47,7 +47,7 @@
     "@celo/client": "f7095b7",
     "@celo/walletkit": "^0.0.4",
     "@celo/react-native-sms-retriever": "git+https://github.com/celo-org/react-native-sms-retriever#d3a2fdb",
-    "@celo/utils": "^0.0.5",
+    "@celo/utils": "^0.0.6-beta3",
```
Comment (ashishb) [inline packages/mobile/package.json:50]: Unrelated?

File: packages/protocol/package.json [hunk 32]
```diff
@@ -70,7 +68,7 @@
   "devDependencies": {
     "@celo/ganache-cli": "git+https://github.com/celo-org/ganache-cli.git#2ebdf9f",
     "@celo/typescript": "0.0.1",
-    "@celo/utils": "^0.0.5",
+    "@celo/utils": "^0.0.6-beta3",
```
Comment (ashishb) [inline packages/protocol/package.json:71]: Unrelated?

File: packages/verification-pool-api/package.json [hunk 33]
```diff
@@ -50,7 +50,6 @@
     "minimist": "^1.2.0",
     "supertest": "^3.4.2",
     "ts-jest": "^24.0.0",
-    "ts-node": "^7.0.1",
```
Comment (ashishb) [inline packages/verification-pool-api/package.json:53]: Can you send a separate PR for this?

File: packages/contractkit/src/wrappers/Validators.ts (hunk 28 above)
Comment (mcortesi) [reply packages/contractkit/src/wrappers/Validators.ts:116]: it has one

File: packages/contractkit/test/driver.test.ts (hunk 23 above)
Comment (mcortesi) [reply packages/contractkit/test/driver.test.ts]: a user might want to connect to it's own node...

File: packages/contractkit/test/driver.test.ts (hunk 29 above)
Comment (mcortesi) [reply packages/contractkit/test/driver.test.ts]: ignore this file... it's just an example and it would be removed later

File: packages/contractkit/src/address-registry.ts (hunk 14 above)
Comment (asaj) [reply packages/contractkit/src/address-registry.ts:11]: It's possible, registry pointers can change. Again, not something we need to address for now

File: packages/contractkit/src/kit.ts (hunk 16 above)
Comment (asaj) [reply packages/contractkit/src/kit.ts]: 🤷‍♂

Comment (asaj) [reply packages/contractkit/src/kit.ts]: I think @ashishb is our expert on locally signed txs

File: packages/contractkit/src/wrappers/GoldTokenWrapper.ts (hunk 21 above)
Comment (asaj) [reply packages/contractkit/src/wrappers/GoldTokenWrapper.ts]: I believe there is an issue

File: packages/contractkit/test/driver.test.ts (hunk 23 above)
Comment (asaj) [reply packages/contractkit/test/driver.test.ts]: Should we remove this file then?

File: packages/contractkit/src/kit.ts [hunk 34]
```diff
@@ -0,0 +1,70 @@
...
+  }
+
+  sendTransaction(tx: Tx) {
+    const promiEvent = this.web3.eth.sendTransaction({
+      from: this.defaultOptions.from,
+      gasPrice: '0',
+      // @ts-ignore
+      gasCurrency: this.defaultOptions.gasCurrency,
+      // @ts-ignore
+      gasFeeRecipient: this.defaultOptions.gasFeeRecipient,
```
Comment (asaj) [inline packages/contractkit/src/kit.ts]: I don't think we should be specifying this explicitly, and instead we should let `celo-blockchain` set this for us

File: packages/contractkit/src/utils/send-tx.ts [hunk 35]
```diff
@@ -0,0 +1,50 @@
...
+    debug('estimatedGas: %s', gas)
+    // logger(EstimatedGas(estimatedGas))
+  }
+
+  const promiEvent = tx.send({
+    from: txOptions.from,
+    // @ts-ignore
+    gasCurrency: txOptions.gasCurrency,
+    // @ts-ignore
+    gasFeeRecipient: txOptions.gasFeeRecipient,
```
Comment (asaj) [inline packages/contractkit/src/utils/send-tx.ts]: Ditto

File: packages/contractkit/test/driver.test.ts (hunk 23 above)
Comment (asaj) [inline packages/contractkit/test/driver.test.ts]: Should we delete this file?

File: packages/contractkit/package.json (hunk 9 above)
Comment (mcortesi) [reply packages/contractkit/package.json:1]: done

File: packages/contractkit/src/kit.ts (hunk 17 above)
Comment (mcortesi) [reply packages/contractkit/src/kit.ts:52]: I'll add a comment for now

File: packages/dappkit/package.json (hunk 30 above)
Comment (mcortesi) [reply packages/dappkit/package.json]: I've added to "Other Changes" section in the PR.
It was necessary to fix this

File: packages/mobile/package.json (hunk 31 above)
Comment (mcortesi) [reply packages/mobile/package.json:50]: Same

File: packages/protocol/package.json (hunk 32 above)
Comment (mcortesi) [reply packages/protocol/package.json:71]: same

File: packages/contractkit/test/driver.test.ts (hunk 23 above)
Comment (mcortesi) [reply packages/contractkit/test/driver.test.ts]: will do

File: packages/contractkit/src/utils/send-tx.ts (hunk 35 above)
Comment (mcortesi) [reply packages/contractkit/src/utils/send-tx.ts]: done

File: packages/contractkit/src/kit.ts (hunk 34 above)
Comment (mcortesi) [reply packages/contractkit/src/kit.ts]: removed with comment

File: packages/verification-pool-api/package.json (hunk 33 above)
Comment (mcortesi) [reply packages/verification-pool-api/package.json:53]: added to Other Changes

File: packages/contractkit/src/kit.ts (hunk 16 above)
Comment (mcortesi) [reply packages/contractkit/src/kit.ts]: Removed and left a comment

File: packages/contractkit/src/address-registry.ts (hunk 14 above)
Comment (mcortesi) [reply packages/contractkit/src/address-registry.ts:11]: Ok
Added and issue for this https://github.com/celo-org/celo-monorepo/issues/575

File: packages/contractkit/src/utils/send-tx.ts [hunk 36]
```diff
@@ -39,11 +39,12 @@ export async function sendTransaction<T>(
     from: txOptions.from,
     // @ts-ignore
     gasCurrency: txOptions.gasCurrency,
-    // @ts-ignore
-    gasFeeRecipient: txOptions.gasFeeRecipient,
+    // TODO needed for locally signed tx, ignored by now (celo-blockchain with set it)
```
Comment (asaj) [inline packages/contractkit/src/utils/send-tx.ts:42]: nit: s/with/will

**Review by asaj — DISMISSED** [review]

**Review by asaj — APPROVED** [review]
